lib/core/src/Cardano/Wallet.hs

```diff
@@ -1484,44 +1494,65 @@ instance Buildable PartialTx where
 balanceTransaction
     :: forall m s k ctx.
         ( HasTransactionLayer k ctx
+        , GenChange s
+        , MonadRandom m
         , HasLogger m WalletWorkerLog ctx
+        )
+    => ctx
+    -> ArgGenChange s
+    -> (W.ProtocolParameters, Cardano.ProtocolParameters)
+    -> TimeInterpreter (Either PastHorizonException)
+    -> (UTxOIndex WalletUTxO, Wallet s, Set Tx)
+    -> PartialTx
+    -> ExceptT ErrBalanceTx m SealedTx
+balanceTransaction ctx change pp ti wallet ptx = do
+    let balanceWith strategy = balanceTransactionWithSelectionStrategy @m @s @k
+            ctx change pp ti wallet strategy ptx
+    balanceWith SelectionStrategyOptimal
+        `catchE` \case
+            ErrBalanceTxMaxSizeLimitExceeded
+                -> balanceWith SelectionStrategyMinimal
+            otherErr
+                -> throwE otherErr
+
+balanceTransactionWithSelectionStrategy
+    :: forall m s k ctx.
+        ( HasTransactionLayer k ctx
         , GenChange s
         , MonadRandom m
+        , HasLogger m WalletWorkerLog ctx
         )
     => ctx
     -> ArgGenChange s
     -> (W.ProtocolParameters, Cardano.ProtocolParameters)
     -> TimeInterpreter (Either PastHorizonException)
     -> (UTxOIndex WalletUTxO, Wallet s, Set Tx)
+    -> SelectionStrategy
     -> PartialTx
     -> ExceptT ErrBalanceTx m SealedTx
-balanceTransaction
+balanceTransactionWithSelectionStrategy
     ctx
     generateChange
     (pp, nodePParams)
     ti
-    (internalUtxoAvailable, wallet, pendingTxs)
-    (PartialTx partialTx@(cardanoTx -> Cardano.InAnyCardanoEra _ (Cardano.Tx (Cardano.TxBody bod) _)) externalInputs redeemers)
+    (internalUtxoAvailable, wallet, _pendingTxs)
+    selectionStrategy
+    ptx@(PartialTx partialTx@(cardanoTx -> Cardano.InAnyCardanoEra _ (Cardano.Tx (Cardano.TxBody bod) _)) externalInputs redeemers)
```

Comment:
This could benefit from some wrapping, as:
- it's hard to see where the nesting is
- it's a very long line

I've made a commit here which attempts to fix the issue:
3b5c5f83536ddc0cdefc7e997dbd173f63dfb646